File: src/repotask/workflow/store.py
"""Artifact store for work in progress: `.repo-task/work/<ticket>/`.

The CLI owns these files so each workflow step has a durable input for the next
one, and so a new agent session can pick the work up without re-fetching.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from repotask.config.models import RepoTaskConfig
from repotask.errors import RepoTaskError
from repotask.files import read_optional, write_text

SOURCE = "source.md"
SUMMARY = "summary.md"
ANALYSIS = "analysis.json"
PLAN = "plan.json"
META = "meta.json"


def now() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="seconds")


def safe_id(ticket_id: str) -> str:
    """Ticket ids become directory names, so keep them boring.

    Leading and trailing dots are stripped as well as separators, so nothing that
    reduces to a relative path fragment survives.
    """
    cleaned = re.sub(r"[^A-Za-z0-9._-]+", "-", ticket_id).strip("-.")
    if not cleaned:
        raise RepoTaskError(f"Invalid ticket id: {ticket_id!r}")
    return cleaned


@dataclass(frozen=True)
class WorkItem:
    root: Path
    ticket_id: str

    @property
    def source_path(self) -> Path:
        return self.root / SOURCE

    def read(self, name: str) -> str | None:
        return read_optional(self.root / name)

    def read_json(self, name: str) -> dict[str, Any]:
        text = self.read(name)
        return json.loads(text) if text else {}

    def write(self, name: str, content: str) -> str:
        write_text(self.root / name, content if content.endswith("\n") else content + "\n")
        return str(self.root / name)

    def write_json(self, name: str, data: dict[str, Any]) -> str:
        return self.write(name, json.dumps(data, indent=2))

    def touch_meta(self, **fields: Any) -> None:
        meta = self.read_json(META)
        meta.update(fields)
        meta.setdefault("ticketId", self.ticket_id)
        meta.setdefault("createdAt", now())
        meta["updatedAt"] = now()
        self.write_json(META, meta)

    def require(self, name: str, hint: str) -> str:
        content = self.read(name)
        if content is None:
            raise RepoTaskError(f"{self.root / name} not found. Run `{hint}` first.")
        return content


def work_item(config: RepoTaskConfig, ticket_id: str) -> WorkItem:
    return WorkItem(root=config.work_dir / safe_id(ticket_id), ticket_id=ticket_id)


def list_work(config: RepoTaskConfig) -> list[WorkItem]:
    if not config.work_dir.is_dir():
        return []
    return [
        WorkItem(root=path, ticket_id=path.name)
        for path in sorted(config.work_dir.iterdir())
        if path.is_dir()
    ]

File: src/workflow/store.rs
//! Artifact store for work in progress: `.repo-task/work/<ticket>/`.
//!
//! The CLI owns these files so each workflow step has a durable input for the next
//! one, and so a new agent session can pick the work up without re-fetching.

use std::path::PathBuf;

use anyhow::{bail, Result};
use serde_json::{json, Value};

use crate::config::RepoTaskConfig;
use crate::kb::source::now_iso;

pub const SOURCE: &str = "source.md";
pub const SUMMARY: &str = "summary.md";
pub const ANALYSIS: &str = "analysis.json";
pub const PLAN: &str = "plan.json";
pub const META: &str = "meta.json";

/// Ticket ids become directory names, so keep them boring. Leading and trailing dots
/// are stripped as well as separators, so nothing that reduces to a relative path
/// fragment survives.
pub fn safe_id(ticket_id: &str) -> Result<String> {
    let mapped: String = ticket_id
        .chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || c == '.' || c == '_' || c == '-' {
                c
            } else {
                '-'
            }
        })
        .collect();
    let cleaned = mapped.trim_matches(['-', '.']).to_string();
    if cleaned.is_empty() {
        bail!("Invalid ticket id: {ticket_id:?}");
    }
    Ok(cleaned)
}

pub struct WorkItem {
    pub root: PathBuf,
    pub ticket_id: String,
}

impl WorkItem {
    pub fn source_path(&self) -> PathBuf {
        self.root.join(SOURCE)
    }

    pub fn read(&self, name: &str) -> Option<String> {
        std::fs::read_to_string(self.root.join(name)).ok()
    }

    pub fn read_json(&self, name: &str) -> Value {
        self.read(name)
            .and_then(|text| serde_json::from_str(&text).ok())
            .unwrap_or_else(|| json!({}))
    }

    pub fn write(&self, name: &str, content: &str) -> Result<String> {
        std::fs::create_dir_all(&self.root)?;
        let path = self.root.join(name);
        let content = if content.ends_with('\n') {
            content.to_string()
        } else {
            format!("{content}\n")
        };
        std::fs::write(&path, content)?;
        Ok(path.to_string_lossy().to_string())
    }

    pub fn write_json(&self, name: &str, data: &Value) -> Result<String> {
        self.write(name, &serde_json::to_string_pretty(data)?)
    }

    pub fn touch_meta(&self, fields: &[(&str, Value)]) -> Result<()> {
        let mut meta = self.read_json(META);
        let map = meta
            .as_object_mut()
            .expect("read_json always yields an object");
        for (key, value) in fields {
            map.insert((*key).to_string(), value.clone());
        }
        map.entry("ticketId")
            .or_insert_with(|| json!(self.ticket_id));
        map.entry("createdAt").or_insert_with(|| json!(now_iso()));
        map.insert("updatedAt".into(), json!(now_iso()));
        self.write_json(META, &meta)?;
        Ok(())
    }

    pub fn require(&self, name: &str, hint: &str) -> Result<String> {
        match self.read(name) {
            Some(content) => Ok(content),
            None => bail!(
                "{} not found. Run `{hint}` first.",
                self.root.join(name).display()
            ),
        }
    }
}

pub fn work_item(config: &RepoTaskConfig, ticket_id: &str) -> Result<WorkItem> {
    Ok(WorkItem {
        root: config.work_dir().join(safe_id(ticket_id)?),
        ticket_id: ticket_id.to_string(),
    })
}

#[cfg(test)]
mod tests {
    use super::safe_id;

    #[test]
    fn keeps_ordinary_ticket_ids() {
        assert_eq!(safe_id("ACME-12").unwrap(), "ACME-12");
        assert_eq!(safe_id("bug_42.1").unwrap(), "bug_42.1");
    }

    #[test]
    fn replaces_separators() {
        assert_eq!(safe_id("feature/ACME 12").unwrap(), "feature-ACME-12");
    }

    #[test]
    fn rejects_path_traversal() {
        assert!(safe_id("../..").is_err());
        assert!(safe_id("..").is_err());
        assert!(safe_id("/").is_err());
        assert!(safe_id("").is_err());
    }
}
